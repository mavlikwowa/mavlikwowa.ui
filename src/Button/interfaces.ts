/* Libraries */
import React from 'react';

/* Sizes of button */
export enum ButtonSize {
  SMALL = 'SMALL',
  MEDIUM = 'MEDIUM',
  LARGE = 'LARGE',
}

/* Props of Button component */
export interface ButtonProps extends React.HTMLProps<HTMLButtonElement> {
  /* Settings of width (in spite of the size settings, changes the button`s width) */
  width?: string;
  /* A flag of loading (shows a preloader) */
  isLoading?: boolean;
  /* Settings of Background */
  background?: string;
  /* Size */
  buttonSize?: ButtonSize;
  /* Color */
  borderColor?: string;
  /* A showing element */
  // children: React.ReactNode;
}
