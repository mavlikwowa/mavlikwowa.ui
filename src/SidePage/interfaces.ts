/* Libraries */
import React from 'react';

export interface SidePageProps
  extends React.HTMLProps<HTMLDivElement>,
    SidePageContainerProps {
  /* a header`s element */
  header?: React.ReactNode;
  /* An action after click on Close Icon */
  onCloseClick: () => void;
}

export interface SidePageContainerProps {
  /* a showing flag */
  show: boolean;
  /* a color of background */
  background?: string;
}
