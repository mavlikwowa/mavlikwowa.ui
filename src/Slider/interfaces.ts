export interface SliderProps {
  data: React.ReactNode[];
  width?: string;
  height?: string;
}

export interface StyledSliderContentProps {
  slideAmount: number;
  // Important! We should calculate a content width
  width: number;
  height?: string;
}

export interface StyledSliderContainerProps {
  width?: string;
  height?: string;
}
