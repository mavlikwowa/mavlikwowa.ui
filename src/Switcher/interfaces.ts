export interface SwitcherProps extends React.HTMLProps<HTMLSpanElement> {
  hasChildren?: React.ReactNode;
  checked?: boolean;
}

export interface StyledSwitcherProps {
  hasChildren: boolean;
}
