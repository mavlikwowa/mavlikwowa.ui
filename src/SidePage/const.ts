// amount of during animation
export const ANIMATION_DURATION = 1000;
