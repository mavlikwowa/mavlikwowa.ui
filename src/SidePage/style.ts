/* Libraries */
import styled from '@emotion/styled';

/* Interfaces */
import { SidePageContainerProps } from './interfaces';

/* Constants */
import { ANIMATION_DURATION } from './const';

/* A styled container */
export const SidePageContainer = styled.div<SidePageContainerProps>`
  ${({ show, background }: SidePageContainerProps) => {
    return `
      display: ${show ? 'block' : 'none'};
      position: fixed;
      top: 0;
      right: 0;
      height: 100%;
      background: ${background || 'white'};
      width: 30%;
      z-index: 999;
      padding: 32px 32px 32px 24px;
      box-sizing: border-box;
      box-shadow: 0 0 10px rgba(0,0,0,0.5);
      transform: translateX(100%);
      transition: transform ${ANIMATION_DURATION}ms;
      
      @media (max-width: 940px) {
        width: 100%;
      }
      &.open {
        transform: translateX(0);
        transition: transform ${ANIMATION_DURATION}ms;
      }
    `;
  }}
`;

/* A styled header */
export const SidePageHeader = styled.div`
  ${() => {
    return `
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: space-between;
      margin-bottom: 24px;
      
      > span: last-of-type {
        outline: none;
        cursor: pointer;
        transition: transform 0.5s;
        &:hover {
          transform: rotate(90deg);  
        }  
      }
    `;
  }}
`;
