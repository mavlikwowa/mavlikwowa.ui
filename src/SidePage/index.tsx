/* Libraries */
import React, { useRef, RefObject, useEffect } from 'react';

/* Interfaces */
import { SidePageProps } from './interfaces';

/* Styled components */
import { SidePageContainer, SidePageHeader } from './style';

/* Components */
import CloseIcon from '../Icons/CloseIcon';

/* Constants */
import { ANIMATION_DURATION } from './const';

const SidePage: React.FC<SidePageProps> = ({
  show,
  background,
  children,
  header,
  onCloseClick,
}: SidePageProps) => {
  // A ref of container. Using for an animation
  const containerRef: RefObject<HTMLDivElement> | null = useRef(null);
  // Adds a class after component is showed
  useEffect(() => {
    if (!containerRef) return;
    if (containerRef?.current) containerRef.current.classList.add('open');
  }, [show]);
  // Closes the component after click
  const closeComponent = () => {
    if (!containerRef) return;
    if (containerRef?.current) containerRef.current.classList.remove('open');
    setTimeout(onCloseClick, ANIMATION_DURATION);
  };
  return (
    <SidePageContainer ref={containerRef} show={show} background={background}>
      <SidePageHeader>
        {header}
        <span role="none" onClick={closeComponent}>
          <CloseIcon />
        </span>
      </SidePageHeader>
      {children}
    </SidePageContainer>
  );
};

export default SidePage;
