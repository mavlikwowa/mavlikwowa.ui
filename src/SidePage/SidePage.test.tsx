/* Libraries */
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

/* Components */
import SidePage from './index';
import CloseIcon from '../Icons/CloseIcon';

/* Interfaces */
import { SidePageProps } from './interfaces';

/* Constants */
import { ANIMATION_DURATION } from './const';

/* A mock for onCloseClick */
const mockCloseClick = jest.fn();

/** Setups */
const setUp = () => {
  /* A test`s data */
  const props: SidePageProps = {
    show: true,
    onCloseClick: mockCloseClick,
    children: 'SidePage text',
  };
  /* Component */
  const wrapper = shallow(<SidePage {...props}>{props.children}</SidePage>);
  return { props, wrapper };
};

/* Tests */
describe('Test component SidePage', () => {
  let wrapper: ShallowWrapper;
  let props: SidePageProps;

  beforeEach(() => {
    ({ wrapper, props } = setUp());
  });

  afterEach(() => {
    if (!wrapper) return;
    wrapper.unmount();
  });

  it('should render a component', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should find a child text', () => {
    expect(wrapper.childAt(1).text()).toBe(props.children);
  });

  it('should call a mock function', () => {
    wrapper.find(CloseIcon).simulate('click');
    setTimeout(() => {
      expect(mockCloseClick).toHaveBeenCalled();
    }, ANIMATION_DURATION);
  });
});
