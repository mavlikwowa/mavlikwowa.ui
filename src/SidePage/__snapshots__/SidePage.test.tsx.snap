// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component SidePage should render a component 1`] = `
<Styled(div)
  show={true}
>
  <Styled(div)>
    <span
      onClick={[Function]}
      role="none"
    >
      <CloseIcon />
    </span>
  </Styled(div)>
  SidePage text
</Styled(div)>
`;
